codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/DocumentationConverter.java

```diff
@@ -119,33 +134,102 @@ public void head(Node node, int depth) {
+            // check the last line's remaining space to see if test should be
+            // written to current line or new line
+            // note that wrapped text will not contain desired indent at the beginning,
+            // so indent will be added to the wrapped text if it is written to a new line
+
+            // if the last line text has reached docWrapLength, directly write at new line
+            if (lastLineRemaining <= 0) {
+                writeNewline();
+                text = StringUtils.wrap(text, docWrapLength, newLineIndent, false);
+                writeInLine(StringUtils.stripStart(newLineIndent, "\n") + text,
+                        docWrapLength - (text.length() - 1 - text.lastIndexOf(newLineIndent)));
+                return;
             }
+
+            // if the last line remaining space is enough for text, just write it to the
+            // current line
+            if (lastLineRemaining >= text.length()) {
+                ensureSplit(' ', text);
+                writeInLine(text, lastLineRemaining - text.length());
+                return;
+            }
+
+            // if the last line remaining space is not enough for the whole text, try to cut
+            // prefix text up to remaining
+            // spaces length and append to the current line, then write remaining suffix
+            // text to new line
+            int lastSpace = text.substring(0, lastLineRemaining).lastIndexOf(" ");
+            if (lastSpace != -1) {
+                String appendString = text.substring(0, lastSpace + 1);
+                ensureSplit(' ', appendString);
+                writer.writeInline(appendString);
+                text = StringUtils.wrap(text.substring(appendString.length()), docWrapLength, newLineIndent, false);
+            } else {
+                text = StringUtils.wrap(text, docWrapLength, newLineIndent, false);
+            }
+
+            writeNewline();
+            writeInLine(StringUtils.stripStart(newLineIndent, "\n") + text,
+                    docWrapLength - (text.length() - 1 - text.lastIndexOf(newLineIndent)));
+        }
+
+        private void ensureSplit(char split, String text) {
+            if (text.charAt(0) != split && lastString != null && !lastString.isEmpty()
+                    && lastString.charAt(lastString.length() - 1) != split) {
+                writeInLine(split + "", lastLineRemaining - 1);
+            }
+        }
+
+        private void writeNewline() {
+            // While jsoup will strip out redundant whitespace, it will still leave some. If
+            // we
+            // start a new line then we want to make sure we don't keep any prefixing
+            // whitespace.
+            // need to refresh last string written and last line remaining white space to
+            // reach docWrapLength
+            shouldStripPrefixWhitespace = true;
+            writer.write("");
+            lastString = null;
```

Comment:
You'll probably have an easier time elsewhere if you use the empty string instead of `null` to represent no text.